src/gsi/read_gsb.f90

```diff
@@ -0,0 +1,561 @@
+  use qcmod, only: njqc
+  use deter_sfc_mod, only: deter_sfc2
+  use gridmod, only: nlon,nlat,rlats,rlons,regional,fv3_regional,diagnostic_reg,& 
+                     rotate_wind_ll2xy,rotate_wind_xy2ll,tll2xy,txy2ll
+  use gsi_4dvar, only: time_4dvar, iwinbgn, l4dvar,l4densvar,winlen
+  use gsi_io, only: verbose
+  use mpimod, only: npe
+  use obsmod, only: bmiss,perturb_obs,perturb_fact,ran01dom
+  implicit none
+
+! Declare passed variables
+  character(len=*)                      ,intent(in   ) :: infile,obstype
+  character(len=20)                     ,intent(in   ) :: sis
+  integer(i_kind)                       ,intent(in   ) :: lunout,nrec_start
+  integer(i_kind)                       ,intent(inout) :: nread,ndata,nodata
+  integer(i_kind),dimension(npe)        ,intent(inout) :: nobs
+  real(r_kind)                          ,intent(in   ) :: gstime
+  real(r_kind)                          ,intent(in   ) :: twindin
+
+! Declare local parameters
+  real(r_kind),parameter:: one_minute = 0.01666667_r_kind
+  real(r_kind),parameter:: minus_one_minute=-0.01666667_r_kind
+  real(r_kind),parameter:: r90  = 90.0_r_kind
+  real(r_kind),parameter:: r360 = 360.0_r_kind
+  real(r_kind),parameter:: r2000 = 2000.0_r_kind
+  real(r_kind),parameter:: emerr= 0.2_r_kind
+
+! Declare local variables
+  character(8) :: subset
+  character(80) :: hdstr, obstr
+
+  integer(i_kind) :: ireadsb,ireadmg, ithin
+  integer(i_kind) :: lunin, nc, ncsave
+  integer(i_kind) :: nmsgmax, mxtb, maxobs, nmsg
+  integer(i_kind) :: i, k, kl, kx, k1, k2, nreal, ntread, ntmatch 
+  integer(i_kind) :: iobsub
+  integer(i_kind) :: ncount_ps, ncount_q, ncount_t, ncount_uv
+  integer(i_kind) :: irec, iret, ncx, levs, iout, ntest, nvtest
+  integer(i_kind) :: idate, nmind, ilat, ilon, nchanl, idomsfc
+  integer(i_kind),dimension(5):: idate5
+  integer(i_kind),dimension(nconvtype)::ntxall
+  integer(i_kind),dimension(nconvtype+1)::ntx
+  integer(i_kind),allocatable,dimension(:,:):: tab
+  integer(i_kind),allocatable,dimension(:):: nrep
+
+  logical :: tob,qob,uvob
+  logical :: outside      
+  logical :: print_verbose
+  
+  real(r_kind) :: dlon, dlat, dlat_earth_deg, dlon_earth_deg, dlat_earth, dlon_earth
+  real(r_kind) :: rlon00, rlat00, cdist, disterr, disterrmax, vdisterrmax, dlnpob
+  real(r_kind) :: toff, t4dv, tdiff
+  real(r_kind) :: uwind, vwind, u0, v0, u00, v00, ppb, usage
+  real(r_kind) :: obserr, var_jb, del, ediff
+  real(r_kind) :: tsavg,ff10,sfcr,zz
+  real(r_kind) :: qjbmin,tjbmin,wjbmin
+  real(r_kind) :: terrmin=half
+  real(r_kind) :: werrmin=one
+  real(r_kind) :: qerrmin=0.05_r_kind
+  real(r_kind),allocatable,dimension(:,:):: cdata_all   !,cdata_out
+  real(r_double) :: rstation_id
+  real(r_double),dimension(9,1):: hdr
+  real(r_double),dimension(7,1):: obsdat
+
+! data statements
+  data hdstr  /'YEAR MNTH DAYS HOUR MINU SECO CLATH CLONH LSTN'/
+  data obstr  /'PRLC HGHT TMDB REHU SPFH WDIR WSPD' /
+  !data levstr  /'PRLC'/   !!!!! Check
```

Comment:
The `!!!!! Check` is interesting.  Is there something we need to check now or in the future?